File: app/src/main/java/com/ardev/builder/compiler/Builder.java
package com.ardev.builder.compiler;

import androidx.annotation.MainThread;



import com.ardev.builder.exception.CompilationFailedException;
import com.ardev.builder.project.api.Module;
import java.io.IOException;
import java.util.List;

public interface Builder<T extends Module> {

    interface TaskListener {
        @MainThread
        void onTaskStarted(String name, String message, int progress);
    }

    void setTaskListener(TaskListener taskListener);

    T getModule();

    void build(BuildType type) throws CompilationFailedException, IOException;

    

    List<Task<? super T>> getTasks(BuildType type);
}

File: app/src/main/java/com/ardev/builder/compiler/Builder.kt
package com.ardev.builder.compiler

import androidx.annotation.MainThread
import com.ardev.builder.exception.CompilationFailure
import com.ardev.builder.project.api.Module
import java.io.IOException
import java.util.List

interface Builder<T : Module> {

    interface TaskListener {
        @MainThread
        fun onTaskStarted(name: String, message: String, progress: Int)
    }

    fun setTaskListener(taskListener: TaskListener)

    val module: T

    @Throws(CompilationFailure::class, IOException::class)
    fun build(type: BuildType)

    fun getTasks(type: BuildType): List<Task<in T>>
}
